diff --git a/contrib/minizip/Makefile.am b/contrib/minizip/Makefile.am
index d343011eb..4f4661023 100644
--- a/contrib/minizip/Makefile.am
+++ b/contrib/minizip/Makefile.am
@@ -26,13 +26,15 @@ libminizip_la_LDFLAGS = $(AM_LDFLAGS) -version-info 1:0:0 -lz
 
 minizip_includedir = $(includedir)/minizip
 minizip_include_HEADERS = \
-	crypt.h \
 	ioapi.h \
 	mztools.h \
 	unzip.h \
 	zip.h \
 	${iowin32_h}
 
+noinst_HEADERS = \
+	crypt.h
+
 pkgconfigdir = $(libdir)/pkgconfig
 pkgconfig_DATA = minizip.pc
 
