diff -Npur "ndiswrapper-1 .59.orig/driver/loader.c" "ndiswrapper-1 .59/driver/loader.c"
--- "ndiswrapper-1 .59.orig/driver/loader.c"	2013-11-28 20:42:35.000000000 +0100
+++ "ndiswrapper-1 .59/driver/loader.c"	2015-08-01 18:25:08.302141265 +0200
@@ -22,6 +22,7 @@
 #include <linux/kmod.h>
 #include <linux/miscdevice.h>
 #include <asm/uaccess.h>
+#include <linux/vmalloc.h>
 
 /*
   Network adapter: ClassGuid = {4d36e972-e325-11ce-bfc1-08002be10318}
diff -Npur "ndiswrapper-1 .59.orig/driver/ntoskernel.c" "ndiswrapper-1 .59/driver/ntoskernel.c"
--- "ndiswrapper-1 .59.orig/driver/ntoskernel.c"	2013-11-28 20:42:36.000000000 +0100
+++ "ndiswrapper-1 .59/driver/ntoskernel.c"	2015-08-01 18:24:55.895942032 +0200
@@ -19,6 +19,7 @@
 #include "pnp.h"
 #include "loader.h"
 #include "ntoskernel_exports.h"
+#include <linux/vmalloc.h>
 
 /* MDLs describe a range of virtual address with an array of physical
  * pages right after the header. For different ranges of virtual
diff -Npur "ndiswrapper-1 .59.orig/driver/pe_linker.c" "ndiswrapper-1 .59/driver/pe_linker.c"
--- "ndiswrapper-1 .59.orig/driver/pe_linker.c"	2013-11-28 20:42:36.000000000 +0100
+++ "ndiswrapper-1 .59/driver/pe_linker.c"	2015-08-01 18:28:37.992752176 +0200
@@ -25,6 +25,7 @@
 //#define DEBUGLINKER 2
 
 #include "ntoskernel.h"
+#include <linux/vmalloc.h>
 
 #endif
 
