diff -Nuar a/src/init.c b/src/init.c
--- a/src/init.c	2021-02-21 21:02:25.000000000 +0300
+++ b/src/init.c	2021-03-04 21:59:54.200890526 +0300
@@ -1492,7 +1492,7 @@
 			for(old = newFamily; old; old = old->next)
 				if (strpbrk(old->rlevel, "S"))  break;
 			if (old == NULL)
-				snprintf(buf, sizeof(buf), "~~:S:wait:%s\n", SULOGIN);
+				snprintf(buf, sizeof(buf), "~~:S:wait:%s -l\n", SHELL);
 			else
 				continue;
 		}
