diff --git a/pisi/actionsapi/kerneltools.py b/pisi/actionsapi/kerneltools.py
index df3de3b..2560b47 100644
--- a/pisi/actionsapi/kerneltools.py
+++ b/pisi/actionsapi/kerneltools.py
@@ -253,6 +253,9 @@ def installLibcHeaders(excludes=None):
     # Create directories
     shelltools.makedirs(headers_tmp)
     shelltools.makedirs(headers_dir)
+
+    # Clean up the source directory
+    autotools.make('mrproper')
     
     # install the headers
     autotools.rawInstall(make_cmd, "headers_install")
