diff -Nuar a/pisi/actionsapi/kerneltools.py b/pisi/actionsapi/kerneltools.py
--- a/pisi/actionsapi/kerneltools.py	2020-01-10 04:21:15.000000000 +0300
+++ b/pisi/actionsapi/kerneltools.py	2020-01-12 00:11:45.464966486 +0300
@@ -254,7 +254,11 @@
     shelltools.makedirs(headers_tmp)
     shelltools.makedirs(headers_dir)
     
+    # Clean up the source directory
+    autotools.make('mrproper')
+    
     # install the headers
+    autotools.make("%s defconfig" % make_cmd)
     autotools.rawInstall(make_cmd, "headers_install")
 
     oldwd = os.getcwd()
