diff -Nuar a/pisi/actionsapi/kerneltools.py b/pisi/actionsapi/kerneltools.py
--- a/pisi/actionsapi/kerneltools.py	2014-03-08 00:03:46.000000000 +0200
+++ b/pisi/actionsapi/kerneltools.py	2020-01-05 19:42:38.461427561 +0300
@@ -268,7 +268,9 @@
     #######################Workaround ends here ...
     
     # make defconfig and install the headers
-    autotools.make("%s defconfig" % make_cmd)
+    #autotools.make("%s defconfig" % make_cmd)
+    #kernel-5.4
+    autotools.make("%s mrproper" % make_cmd)
     autotools.rawInstall(make_cmd, "headers_install")
 
     oldwd = os.getcwd()
