--- glibc-2.22-ORIG/debian/debhelper.in/libc-dev.install	2015-12-02 17:43:57.386493285 +0100
+++ glibc-2.22/debian/debhelper.in/libc-dev.install	2015-12-02 17:44:23.986900216 +0100
@@ -8,6 +8,7 @@
 TMPDIR/LIBDIR/libieee.a LIBDIR
 TMPDIR/LIBDIR/libm.a LIBDIR
 TMPDIR/LIBDIR/libmcheck.a LIBDIR
+TMPDIR/LIBDIR/libmvec_nonshared.a LIBDIR
 TMPDIR/LIBDIR/libnsl.a LIBDIR
 TMPDIR/LIBDIR/libpthread.a LIBDIR
 TMPDIR/LIBDIR/libpthread_nonshared.a LIBDIR

