diff --git a/Makefile.sharedlibrary b/Makefile.sharedlibrary
index 2c98de9..452edca 100644
--- a/Makefile.sharedlibrary
+++ b/Makefile.sharedlibrary
@@ -53,6 +53,7 @@ DUKTAPE_SRCDIR = ./src
 #DUKTAPE_SRCDIR = ./src-noline
 
 CC = gcc
+LIBS = -lm
 
 .PHONY: all
 all: libduktape.$(SO_REALNAME_SUFFIX) libduktaped.$(SO_REALNAME_SUFFIX)
@@ -62,12 +63,12 @@ all: libduktape.$(SO_REALNAME_SUFFIX) libduktaped.$(SO_REALNAME_SUFFIX)
 # to $INSTALL_PREFIX/include on installation.
 
 libduktape.$(SO_REALNAME_SUFFIX):
-	$(CC) -shared -fPIC -Wall -Wextra -Os -Wl,$(LD_SONAME_ARG),libduktape.$(SO_SONAME_SUFFIX) \
-		-o $@ $(DUKTAPE_SRCDIR)/duktape.c
+	$(CC) $(CFLAGS) -shared -fPIC -Wall -Wextra -Os $(LDFLAGS) -Wl,$(LD_SONAME_ARG),libduktape.$(SO_SONAME_SUFFIX) \
+		-o $@ $(DUKTAPE_SRCDIR)/duktape.c $(LIBS)
 
 libduktaped.$(SO_REALNAME_SUFFIX):
-	$(CC) -shared -fPIC -g -Wall -Wextra -Os -Wl,$(LD_SONAME_ARG),libduktaped.$(SO_SONAME_SUFFIX) \
-		-o $@ $(DUKTAPE_SRCDIR)/duktape.c
+	$(CC) $(CFLAGS) -shared -fPIC -Wall -Wextra -Os $(LDFLAGS) -Wl,$(LD_SONAME_ARG),libduktaped.$(SO_SONAME_SUFFIX) \
+		-o $@ $(DUKTAPE_SRCDIR)/duktape.c $(LIBS)
 
 # Symlinks depend on platform conventions.
 .PHONY: install
