From eb78197af36bb0fb95493ebf8fce104be6832ec9 Mon Sep 17 00:00:00 2001
From: Chet Ramey <chet.ramey@case.edu>
Date: Mon, 29 Jan 2018 16:04:56 -0500
Subject: [PATCH] Bash-4.4 patch 18

---
 builtins/read.def | 2 +-
 patchlevel.h      | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/builtins/read.def b/builtins/read.def
index 803bea35..b54b3af6 100644
--- a/builtins/read.def
+++ b/builtins/read.def
@@ -610,7 +610,7 @@ read_builtin (list)
 	}
 
       CHECK_ALRM;
-
+      QUIT;		/* in case we didn't call check_signals() */
 #if defined (READLINE)
 	}
 #endif
diff --git a/patchlevel.h b/patchlevel.h
index 98e714da..f0ee56e4 100644
--- a/patchlevel.h
+++ b/patchlevel.h
@@ -25,6 +25,6 @@
    regexp `^#define[ 	]*PATCHLEVEL', since that's what support/mkversion.sh
    looks for to find the patch level (for the sccs version string). */
 
-#define PATCHLEVEL 17
+#define PATCHLEVEL 18
 
 #endif /* _PATCHLEVEL_H_ */
-- 
2.13.6

