From ad8cf343ab5c442e62e0b2606732760787ae735a Mon Sep 17 00:00:00 2001
From: Nick Clifton <nickc@redhat.com>
Date: Sun, 27 Jul 2025 10:53:38 +0100
Subject: [PATCH 01/23] Re-enable development on the 2.45 branch

---
 bfd/development.sh | 2 +-
 bfd/version.m4     | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/bfd/development.sh b/bfd/development.sh
index 8b004da7d52..b8014261bfb 100644
--- a/bfd/development.sh
+++ b/bfd/development.sh
@@ -16,7 +16,7 @@
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
 # Controls whether to enable development-mode features by default.
-development=false
+development=true
 
 # Indicate whether this is a release branch.
 experimental=false
diff --git a/bfd/version.m4 b/bfd/version.m4
index 00d03934d68..848470098f5 100644
--- a/bfd/version.m4
+++ b/bfd/version.m4
@@ -1 +1 @@
-m4_define([BFD_VERSION], [2.45])
+m4_define([BFD_VERSION], [2.45.0])
-- 
2.49.1

