
__GFP_REPEAT is now __GFP_RETRY_MAYFAIL

Signed-off-by: Thomas Backlund <tmb@mageia.org>

--- linux/3rdparty/ndiswrapper/ntoskernel.h.orig	2017-10-15 01:11:56.041892449 +0300
+++ linux/3rdparty/ndiswrapper/ntoskernel.h	2017-10-15 01:11:06.159204294 +0300
@@ -122,7 +122,7 @@ static cpumask_t cpumasks[NR_CPUS];
 
 #define PCI_DMA_ALLOC_COHERENT(pci_dev,size,dma_handle)			\
 	dma_alloc_coherent(&pci_dev->dev,size,dma_handle,		\
-			   GFP_KERNEL | __GFP_REPEAT)
+			   GFP_KERNEL | __GFP_RETRY_MAYFAIL)
 #define PCI_DMA_FREE_COHERENT(pci_dev,size,cpu_addr,dma_handle)		\
 	dma_free_coherent(&pci_dev->dev,size,cpu_addr,dma_handle)
 #define PCI_DMA_MAP_SINGLE(pci_dev,addr,size,direction)		\
