
Adjust for kernel changes.

Signed-off-by: Thomas Backlund <tmb@mageia.org>

--- linux/net/ipv4/netfilter/ipt_IFWLOG.c.orig
+++ linux/net/ipv4/netfilter/ipt_IFWLOG.c
@@ -138,7 +138,7 @@ static unsigned int ipt_IFWLOG_target(st
 {
 	const struct ipt_IFWLOG_info *info = target_param->targinfo;
 
-	ipt_IFWLOG_packet(skb, target_param->in, target_param->out, info);
+	ipt_IFWLOG_packet(skb, xt_in(target_param), xt_out(target_param), info);
 
 	return XT_CONTINUE;
 }
